### PR TITLE
Create get started with syntax analysis

### PR DESCRIPTION
**Pull request by BillWagner**
Create the get started for using the Roslyn Syntax APIs.  This was adapted from [this topic](https://github.com/dotnet/roslyn/wiki/Getting-Started-C%23-Syntax-Analysis) on the roslyn wiki.

Fixes dotnet/Roslyn-api-docs#9

[Internal review link](https://review.docs.microsoft.com/en-us/dotnet/csharp/roslyn-sdk/get-started/syntax-analysis?branch=pr-en-us-4289)

### CODE REVIEW

**Comment by BillWagner** [comment]
This is ready for review. 

/cc @justcla if you want to review.

File: docs/csharp/roslyn-sdk/get-started/syntax-analysis.md [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+title: Get Started with Syntax Analysis (Roslyn APIs)
+description: An introduction to traversing, querying and walking syntax trees.
+keywords: roslyn, analyzer, code fix
+author: billwagner
+ms.author: wiwagn
+ms.date: 01/29/2018
+ms.topic: conceptual
+ms.prod: .net
+ms.devlang: devlang-csharp
+ms.custom: mvc
```
Comment (mairaw) [inline docs/csharp/roslyn-sdk/get-started/syntax-analysis.md]: my understanding of the mvc template is that some of the steps would have their own headers. @JRAlexander is probably more familiar with that.

File: samples/csharp/roslyn-sdk/SyntaxQuickStart/HelloSyntaxTree/Program.cs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+            // </Snippet3>
+
+            // <Snippet4>
+            MemberDeclarationSyntax firstMember = root.Members[0];
+            WriteLine($"The first member is a {firstMember.Kind()}");
+            NamespaceDeclarationSyntax helloWorldDeclaration = firstMember as NamespaceDeclarationSyntax;
+            // </Snippet4>
+
+            // <Snippet5>
+            WriteLine($"There are {helloWorldDeclaration.Members.Count} members declared in this namespace.");
```
Comment (mairaw) [inline samples/csharp/roslyn-sdk/SyntaxQuickStart/HelloSyntaxTree/Program.cs]: should you account for 1 vs 1+ here? On your results, you ended up with the following sentence:
"There are 1 members declared in this namespace."

Not sure if you want "There is 1 member declared in this namespace." and "There are n members declared in this namespace." for n=2 or more.

**Review by mairaw — CHANGES_REQUESTED** [review]
Looks really good @BillWagner. I've created a PR for more trivial changes on your fork. Left some additional comments/questions here too for you to look at.

File: docs/csharp/roslyn-sdk/get-started/syntax-analysis.md [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+---
+
+# Get Started with Syntax Analysis
+
+In this tutorial you explore the **Syntax API**. The syntax API provides access to the data structures that describe a C# or Visual Basic program. These data structures have enough detail that they can fully represent any program of any size. These structures can describe complete programs that compile and run correctly. They can also describe incomplete programs, as you write them, in the editor.
+
+To enable this rich expression, the data structures and APIs that make up the syntax API are necessarily complex. As is traditional, let's start with what the data structure looks like for the typical "Hello World" program:
```
[the remark this answers is not in the document]
Comment (BillWagner) [reply docs/csharp/roslyn-sdk/get-started/syntax-analysis.md]: OK, but I really liked it....

File: docs/csharp/roslyn-sdk/get-started/syntax-analysis.md [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+By navigating this tree structure you can find any statement, expression, token, or bit of whitespace in a code file!
+
+While you can find anything in a code file using the syntax APIs, most scenarios involve examining small snippets of code, or searching for particular statements or fragments. The two examples that follow show typical uses to browse the structure of code, or search for single statements.
+
+## Traversing Trees
+
+You can examine the nodes in a syntax tree in two ways. You can traverse the tree to examine each node. Alternatively, you can query for specific elements or nodes.
+
+> [!IMPORTANT]
+> The following samples require the **.NET Compiler SDK** installed as part of Visual Studio 2017. You can find the .NET Compiler SDK as the last optional component listed under the **Visual Studio extension development** workload. The templates aren't installed without this component.
```
[the remark this answers is not in the document]
Comment (BillWagner) [reply docs/csharp/roslyn-sdk/get-started/syntax-analysis.md]: This isn't available yet for Visual Studio for Mac.

The API Analyzer will run without having this workload installed (including on a Mac). This workload is needed to create analyzers, not run them.

File: samples/csharp/roslyn-sdk/SyntaxQuickStart/HelloSyntaxTree/Program.cs (hunk 2 above)
Comment (BillWagner) [reply samples/csharp/roslyn-sdk/SyntaxQuickStart/HelloSyntaxTree/Program.cs]: went with singular.

File: docs/csharp/roslyn-sdk/get-started/syntax-analysis.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (mairaw) [reply docs/csharp/roslyn-sdk/get-started/syntax-analysis.md]: It was hard to parse for me. But it's ok if you wanna keep it.

File: docs/csharp/roslyn-sdk/get-started/syntax-analysis.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (mairaw) [reply docs/csharp/roslyn-sdk/get-started/syntax-analysis.md]: Cool! Thanks for the clarification.

File: docs/csharp/roslyn-sdk/get-started/syntax-analysis.md [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+ms.date: 01/29/2018
+ms.topic: conceptual
+ms.prod: .net
+ms.devlang: devlang-csharp
+ms.custom: mvc
+---
+
+# Get Started with Syntax Analysis
+
+In this tutorial you explore the **Syntax API**. The syntax API provides access to the data structures that describe a C# or Visual Basic program. These data structures have enough detail that they can fully represent any program of any size. These structures can describe complete programs that compile and run correctly. They can also describe incomplete programs, as you write them, in the editor.
```
[the remark this answers is not in the document]
Comment (BillWagner) [reply docs/csharp/roslyn-sdk/get-started/syntax-analysis.md]: I capitalized all usages.

File: docs/csharp/roslyn-sdk/get-started/syntax-analysis.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (BillWagner) [reply docs/csharp/roslyn-sdk/get-started/syntax-analysis.md]: I'd rather keep the clarity.

**Review by CamSoper — APPROVED** [review]
The only comment I have is regarding the use of "you."  I know that's frowned upon in the ASP.NET docset, but I'm guessing that's not the case here?

@mairaw - I checked all of your requested changes and he addressed them all.  I think it's good for a merge.

**Comment by CamSoper** [comment]
@BillWagner - Groovy!  Just wanted to double-check.  (Now how do I keep that straight in my brain since I use both repositories?  😆 )

**Comment by mairaw** [comment]
Thanks for double checking @CamSoper. I'm on it now for a second review.

**Comment by mairaw** [comment]
But I'm a bit confused on why this is frowned upon on the asp.net repo.
See here an example from Microsoft style guide: https://docs.microsoft.com/en-us/style-guide/top-10-tips-style-voice#write-like-you-speak

/cc @Rick-Anderson @scottaddie can you guys comment on this? We shouldn't have different rules for our repos.

File: docs/csharp/roslyn-sdk/get-started/syntax-analysis.md [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+
+The sample uses `WriteLine` statements to display information about the syntax trees as they are traversed. You can also learn much more by running the finished program under the debugger. You can examine more of the properties and methods that are part of the syntax tree created for the hello world program.
+
+## Syntax walkers
+
+Often you want to find all nodes of a specific type in a syntax tree, for example, every property declaration in a file. By extending the <xref:Microsoft.CodeAnalysis.CSharp.CSharpSyntaxWalker?displayProperty=nameWithType> class and overriding the <xref:Microsoft.CodeAnalysis.CSharp.CSharpSyntaxVisitor.VisitPropertyDeclaration(Microsoft.CodeAnalysis.CSharp.Syntax.PropertyDeclarationSyntax)> method, you process every property declaration in a syntax tree without knowing its structure beforehand. <xref:Microsoft.CodeAnalysis.CSharp.CSharpSyntaxWalker?displayProperty=nameWithType> is a specific kind of <xref:Microsoft.CodeAnalysis.CSharp.CSharpSyntaxVisitor?displayProperty=nameWithType> that recursively visits a node and each of its children.
```
Comment (mairaw) [inline docs/csharp/roslyn-sdk/get-started/syntax-analysis.md]: nit: do you really want the full name in all the instances?

image

File: docs/csharp/roslyn-sdk/get-started/syntax-analysis.md [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+By navigating this tree structure, you can find any statement, expression, token, or bit of whitespace in a code file.
+
+While you can find anything in a code file using the Syntax APIs, most scenarios involve examining small snippets of code, or searching for particular statements or fragments. The two examples that follow show typical uses to browse the structure of code, or search for single statements.
+
+## Traversing trees
+
+You can examine the nodes in a syntax tree in two ways. You can traverse the tree to examine each node, or you can query for specific elements or nodes.
+
+> [!IMPORTANT]
+> The following samples require the **.NET Compiler SDK** installed as part of Visual Studio 2017. You can find the .NET Compiler SDK as the last optional component listed under the **Visual Studio extension development** workload. The templates aren't installed without this component.
```
Comment (mairaw) [inline docs/csharp/roslyn-sdk/get-started/syntax-analysis.md]: this can be easily missed. Should we add a prerequisites section on top? I was wondering why I was not seeing the template and then went digging for the info in the middle of the doc.

File: docs/csharp/roslyn-sdk/get-started/syntax-analysis.md [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+
+## Syntax walkers
+
+Often you want to find all nodes of a specific type in a syntax tree, for example, every property declaration in a file. By extending the <xref:Microsoft.CodeAnalysis.CSharp.CSharpSyntaxWalker?displayProperty=nameWithType> class and overriding the <xref:Microsoft.CodeAnalysis.CSharp.CSharpSyntaxVisitor.VisitPropertyDeclaration(Microsoft.CodeAnalysis.CSharp.Syntax.PropertyDeclarationSyntax)> method, you process every property declaration in a syntax tree without knowing its structure beforehand. <xref:Microsoft.CodeAnalysis.CSharp.CSharpSyntaxWalker?displayProperty=nameWithType> is a specific kind of <xref:Microsoft.CodeAnalysis.CSharp.CSharpSyntaxVisitor?displayProperty=nameWithType> that recursively visits a node and each of its children.
+
+This example implements a <xref:Microsoft.CodeAnalysis.CSharp.CSharpSyntaxWalker?displayProperty=nameWithType> that examines a syntax tree. It collects `using` directives it finds that aren't importing a `System` namespace.
+
+Create a new C# **Stand-Alone Code Analysis Tool** project; name it "**SyntaxWalker**."
```
Comment (mairaw) [inline docs/csharp/roslyn-sdk/get-started/syntax-analysis.md]: do we need to give the full instructions again or do you think folks would know after creating the first one?

**Review by mairaw — APPROVED** [review]
Left a couple of nit comments. I wanna install the template and follow along the tutorial now to make sure there is nothing missing.

File: docs/csharp/roslyn-sdk/get-started/syntax-analysis.md (hunk 7 above)
Comment (mairaw) [reply docs/csharp/roslyn-sdk/get-started/syntax-analysis.md]: Name also seems to be .NET Compiler Platform SDK, no?

**Comment by scottaddie** [comment]
@mairaw We do still use the word "you", although it's used sparingly. I'm unsure of the reasoning, to be honest.

**Comment by Rick-Anderson** [comment]
Use it when it's more clear. The PU prefers ref style docs. If it doesn't add any clarity, I avoid it.

File: docs/csharp/roslyn-sdk/get-started/syntax-analysis.md (hunk 6 above)
Comment (BillWagner) [reply docs/csharp/roslyn-sdk/get-started/syntax-analysis.md]: I looked at the entire doc and use `nameWithType` on the first use, but only the name after that.

File: docs/csharp/roslyn-sdk/get-started/syntax-analysis.md (hunk 7 above)
Comment (BillWagner) [reply docs/csharp/roslyn-sdk/get-started/syntax-analysis.md]: dotnet/roslyn-api-docs#36 will add a topic to install this template. I'll make sure to reference that new article from all of these when it's written (probably with an include for prereqs.)

File: docs/csharp/roslyn-sdk/get-started/syntax-analysis.md (hunk 8 above)
Comment (BillWagner) [reply docs/csharp/roslyn-sdk/get-started/syntax-analysis.md]: I think this is fine, as it's later in the same article.
